src/Storages/StorageS3.cpp

```diff
@@ -214,7 +215,23 @@ class StorageS3Source::DisclosedGlobIterator::Impl : WithContext
         if (is_finished)
             return {};
 
-        fillInternalBufferAssumeLocked();
+        try
+        {
+            fillInternalBufferAssumeLocked();
+        }
+        catch (...)
+        {
+            /// Iterator is used by several processors from several threads.
```

Comment:
> /// In case of exception it may take some time for threads
             /// to stop processors and they may still use this iterator.
             /// To avoid UB mark the iterator as finished

So we had an exception during adding new files to be iterated and it became invalid and other threads start to use it after we had an exception? Could you add it to the comment? because it is not clear what you mean in your current comment how ub is related to iterator